File: native/windows/Services/WirelessTransferServer.cs
namespace NikonLink.Windows.Services;

public sealed class WirelessTransferServer : IAsyncDisposable
{
    public const int FtpPort = WirelessFtpServer.Port;
    public const int HttpPort = WirelessHttpServer.Port;
    public const string Username = WirelessFtpServer.Username;
    public const string Password = WirelessFtpServer.Password;

    private readonly WirelessFtpServer _ftp;
    private readonly WirelessHttpServer _http;
    private bool _running;

    public WirelessTransferServer(PhotoLibrary library)
    {
        _ftp = new WirelessFtpServer(library);
        _http = new WirelessHttpServer(library);
        _ftp.StatusChanged += (_, status) => PublishStatus(status);
        _http.StatusChanged += (_, status) => PublishStatus(status);
        _ftp.FileReceived += (_, path) => FileReceived?.Invoke(this, path);
        _http.FileReceived += (_, path) => FileReceived?.Invoke(this, path);
        _ftp.Failed += (_, error) => Failed?.Invoke(this, error);
        _http.Failed += (_, error) => Failed?.Invoke(this, error);
    }

    public event EventHandler<string>? StatusChanged;
    public event EventHandler<string>? FileReceived;
    public event EventHandler<Exception>? Failed;

    public bool IsRunning => _running;
    public string LocalAddress => _ftp.LocalAddress;

    public async Task StartAsync()
    {
        if (_running)
        {
            return;
        }
        _running = true;
        try
        {
            await _ftp.StartAsync();
            await _http.StartAsync();
            StatusChanged?.Invoke(
                this,
                "等待 FTP / HTTP / WebDAV 图片");
        }
        catch
        {
            _running = false;
            await _http.StopAsync();
            await _ftp.StopAsync();
            throw;
        }
    }

    public async Task StopAsync()
    {
        if (!_running)
        {
            return;
        }
        _running = false;
        await _http.StopAsync();
        await _ftp.StopAsync();
        StatusChanged?.Invoke(this, "无线收件箱未开启");
    }

    private void PublishStatus(string status)
    {
        if (!_running || status == "无线收件箱未开启")
        {
            return;
        }
        StatusChanged?.Invoke(
            this,
            status == "等待相机无线传输"
                ? "等待 FTP / HTTP / WebDAV 图片"
                : status);
    }

    public async ValueTask DisposeAsync()
    {
        await StopAsync();
        GC.SuppressFinalize(this);
    }
}

File: native/android/app/src/main/java/com/tauber/nikonlink/WirelessTransferServer.java
package com.tauber.nikonlink;

import java.io.File;

final class WirelessTransferServer {
    static final int FTP_PORT = WirelessFtpServer.PORT;
    static final int HTTP_PORT = WirelessHttpServer.PORT;
    static final String USERNAME = WirelessFtpServer.USERNAME;
    static final String PASSWORD = WirelessFtpServer.PASSWORD;

    interface Listener {
        void onStatus(String status);
        void onFileReceived(File file);
        void onError(String message);
    }

    private final WirelessFtpServer ftp;
    private final WirelessHttpServer http;
    private final Listener listener;
    private volatile boolean running;

    WirelessTransferServer(File directory, Listener listener) {
        this.listener = listener;
        ftp = new WirelessFtpServer(
                directory,
                new WirelessFtpServer.Listener() {
                    @Override
                    public void onStatus(String status) {
                        publishProtocolStatus(status);
                    }

                    @Override
                    public void onFileReceived(File file) {
                        WirelessTransferServer.this.listener.onFileReceived(file);
                    }

                    @Override
                    public void onError(String message) {
                        fail("FTP 启动失败：" + message);
                    }
                });
        http = new WirelessHttpServer(
                directory,
                new WirelessHttpServer.Listener() {
                    @Override
                    public void onStatus(String status) {
                        publishProtocolStatus(status);
                    }

                    @Override
                    public void onFileReceived(File file) {
                        WirelessTransferServer.this.listener.onFileReceived(file);
                    }

                    @Override
                    public void onError(String message) {
                        fail("HTTP / WebDAV 启动失败：" + message);
                    }
                });
    }

    synchronized void start() {
        if (running) return;
        running = true;
        ftp.start();
        http.start();
        listener.onStatus("等待 FTP / HTTP / WebDAV 图片");
    }

    synchronized void stop() {
        if (!running) return;
        running = false;
        ftp.stop();
        http.stop();
        listener.onStatus("无线收件箱未开启");
    }

    boolean isRunning() {
        return running;
    }

    String getLocalAddress() {
        return ftp.getLocalAddress();
    }

    private void publishProtocolStatus(String status) {
        if (!running || status.equals("无线收件箱未开启")) return;
        if (status.equals("等待相机无线传输")) {
            listener.onStatus("等待 FTP / HTTP / WebDAV 图片");
        } else {
            listener.onStatus(status);
        }
    }

    private void fail(String message) {
        synchronized (this) {
            if (!running) return;
            running = false;
        }
        ftp.stop();
        http.stop();
        listener.onError(message);
    }
}
